usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -489,6 +519,467 @@ static CK_RV cleanse_attribute(TEMPLATE *template,
+ *
+ * @return true if conditions met
+ *         false otherwise
+ */
+CK_BBOOL ep11tok_pkey_usage_ok(STDLL_TokData_t *tokdata, SESSION *session,
+                               OBJECT *key_obj, CK_MECHANISM *mech)
+{
+    ep11_private_data_t *ep11_data = tokdata->private_data;
+    CK_ATTRIBUTE *opaque_attr = NULL;
+    CK_BBOOL bval = CK_FALSE;
+    CK_RV ret;
+
+    /* Check if CPACF supports the operation implied by this key and mech */
+    if (!pkey_op_supported_by_cpacf(mech))
+        goto done;
+
+    /* Check config option */
+    switch (ep11_data->pkey_mode) {
+    case PKEY_MODE_DISABLED:
+        goto done;
+        break;
+    case PKEY_MODE_DEFAULT:
+    case PKEY_MODE_ENABLE4NONEXTR:
+        /* Use existing pkeys, re-create invalid pkeys, and also create new
+         * pkeys for keys that do not already have one. */
+        if (object_is_extractable(key_obj) ||
+            !object_is_pkey_extractable(key_obj) ||
+            object_is_attr_bound(key_obj) ||
+            !ep11tok_pkey_session_read_write(session) ||
+            !ep11_data->pkey_wrap_supported) {
+            goto done;
+        }
+        if (template_attribute_get_non_empty(key_obj->template,
+                                             CKA_IBM_OPAQUE_PKEY,
+                                             &opaque_attr) != CKR_OK ||
+            !ep11tok_pkey_is_valid(tokdata, key_obj)) {
+            /* this key has either no pkey attr, or it is not valid,
+             * try to create one */
+            ret = ep11tok_pkey_update(tokdata, key_obj);
+            if (ret != CKR_OK) {
+                TRACE_ERROR("error updating the protected key, rc=0x%lx\n", ret);
+                goto done;
+            }
+        }
+        break;
+    default:
+        /* should not occur */
+        TRACE_ERROR("PKEY_MODE %i unsupported.\n", ep11_data->pkey_mode);
+        goto done;
+        break;
+    }
+
+    bval = CK_TRUE;
+
+done:
+
+    return bval;
+}
+
+/**
+ * Wrapper function around ep11tok_pkey_usage_ok for the case where we don't
+ * have a key object. This function is called externally from new_host.c.
+ */
+CK_BBOOL ep11tok_pkey_usage_ok2(STDLL_TokData_t *tokdata, SESSION *session,
+                                CK_OBJECT_HANDLE hkey, CK_MECHANISM *mech)
+{
+    CK_BBOOL success = CK_FALSE;
+    size_t keyblobsize = 0;
+    CK_BYTE *keyblob;
+    OBJECT *key_obj;
+    CK_RV ret;
+
+    ret = h_opaque_2_blob(tokdata, hkey, &keyblob, &keyblobsize, &key_obj,
+                         READ_LOCK);
+    if (ret != CKR_OK) {
+        TRACE_ERROR("%s no blob ret=0x%lx\n", __func__, ret);
+        return CK_FALSE;
+    }
+
+    if (ep11tok_pkey_usage_ok(tokdata, session, key_obj, mech))
+        success = CK_TRUE;
+
+    object_put(tokdata, key_obj, TRUE);
+    key_obj = NULL;
+
+    return success;
+}
+
+/**
+ * This function is called whenever a new secret or private key object is
+ * created. It sets attribute CKA_IBM_PROTKEY_EXTRACTABLE according to the
```

Comment:
Comment: Don't confuse secret or private/public keys with private objects. One means a key type, the other one an attribute of an object. 

I would assume that this function should be called for all key objects, regardless of the key type, and regardless if the object is private or not.

Reply:
This function is indeed called in object_create_skel only if class = secret or private key. It just sets cka_ibm_protkey_extractable to true, depending on cka_extractable of this key, token option, session, etc. But it would not make sense for e.g. public keys or certificates.

Reply:
ok, if we keep the function pkey specific, then this makes sense, but I would still prefer to make it generic (see other comments about this topic). The way it is now it is too much specific for pkey, and has too much internal knowledge about how pkey works in common code.

Reply:
OK, I'm now calling token_specific_set_pkey_attr for all objects and I also added class and mode to its parm list. However, the two parms are currently unused.

Reply:
> However, the two parms are currently unused.

Don't you need to check the class in the implementation in the EP11 token? You want to add the PKEY attribute only for key objects, but not to all objects (although the function is called for all objects). So at least the class parameter should be used, right?

And since this function is now generic (I like that), can we change the name and remove 'pkey' from it s name? I.e. token_specific_set_attrs_for_new_object or something like that?

Reply:
Yes, I'm now checking the class inside the token_specific routine. The routine only considers secret and private keys.